misc: more on values, while loops and better highlighting (#15)
mun-lang/book: src/ch02-04-extern-fn.md

## `extern` functions

> Extern functions are functions defined in Mun without an implementation. They behave exactly the same as regular functions but they have to be provided to the runtime when loading a Mun library. Failure to do so will result in a runtime link error, and loading the library will not work.

Reviewer: ```suggestion
Extern functions are functions defined in Mun without an implementation. They behave exactly the same as regular functions but they have to be provided to the runtime when loading a Mun library. Failure to do so will result in a runtime link error, and loading the library will fail.
```

## `extern` functions

> Extern functions are functions defined in Mun without an implementation. They
> behave exactly the same as regular functions but they have to be provided to the
> runtime when loading a Mun library. Failure to do so will result in a runtime
> link error, and loading the library will fail.

Reviewer: ```suggestion
Extern functions are declared in Mun but their function bodies are defined
externally. They behave exactly the same as regular functions but their
definitions have to be provided to the runtime when loading a Mun library.
Failure to do so will result in a runtime link error, and loading the library
will fail. Take this code for example:
```

> Take this code for example:
> 

Reviewer: ```suggestion
```

```mun
{{#include ../listings/ch02-basic-concepts/listing01.mun}}
```

<span class="caption">Listing 2-1: Random bool in Mun</span>

> The `random` function is marked as an `extern` function which means it must be
> provide to the runtime when loading this library.

Reviewer: ```suggestion
The `random` function is marked as an `extern` function, which means that it
must be provided to the runtime when loading this library.
```

First building the above code as `main.munlib` and then trying to load the
library in Rust using:

```rust
{{#include ../listings/ch02-basic-concepts/listing02.rs}}
```

<span class="caption">Listing 2-2: Load listing 2-1 without adding extern function</span>

will result in an error:

```
Failed to link: function `random` is missing.
```

This indicates that we have to provide the runtime with the `random` method,

> which we can do through the use of the `insert_fn` method. Lets add a method

Reviewer: ```suggestion
which we can do through the use of the `insert_fn` method. Let's add a method
```

that uses the current time as the base of our `random` method:

```rust
{{#include ../listings/ch02-basic-concepts/listing03.rs}}
```

<span class="caption">Listing 2-3: Load listing 2-1 with custom `random` function</span>

Note that we have to explicitly cast the function `random` to `extern "C" fn()
-> i64`. This is because each function in Rust has its own unique type.

> Now if we run this, the error is gone and you should have a function that

Reviewer: ```suggestion
When we run this now, the error is gone and you should have a function that
```

> returns a random boolean written in Mun.

Reviewer: ```suggestion
returns a random boolean in Mun.
```
